### PR TITLE
Few simple code improvements spotted by cppcheck

### PR DESCRIPTION
**Pull request by rouault**

### CODE REVIEW

File: src/operation/union/CoverageUnion.cpp [hunk 1]
```diff
@@ -80,22 +80,21 @@ std::unique_ptr<Geometry> CoverageUnion::polygonize(const GeometryFactory* gf) {
...
+        segment_geoms.emplace_back(std::move(seg_geom));
     }
 
     if (!p.allInputsFormPolygons()) {
         throw geos::util::TopologyException("CoverageUnion cannot process incorrectly noded inputs.");
     }
 
     auto polygons = p.getPolygons();
-    segment_geoms.reset();
+    segment_geoms.clear();
```
Comment (dbaston) [inline src/operation/union/CoverageUnion.cpp:97]: I think this line can be removed; the intent of the `reset` was to free the memory early, but that doesn't seem necessary in the current formulation.

File: src/geomgraph/TopologyLocation.cpp [hunk 2]
```diff
@@ -47,8 +48,9 @@ TopologyLocation::merge(const TopologyLocation& gl)
         location[Position::LEFT] = Location::NONE;
         location[Position::RIGHT] = Location::NONE;
     }
-    for(std::size_t i = 0; i < locationSize; ++i) {
-        if(location[i] == Location::NONE && i < glsz) {
+    const std::size_t maxIndex = std::min(static_cast<std::size_t>(locationSize), glsz);
+    for(std::size_t i = 0; i < maxIndex; ++i) {
```
Comment (dbaston) [inline src/geomgraph/TopologyLocation.cpp:52]: Will need to perf test this; I have noticed enormous and entirely unexpected performance difference from tiny changes to these sorts of small-N loops.

File: src/operation/union/CoverageUnion.cpp (hunk 1 above)
Comment (rouault) [reply src/operation/union/CoverageUnion.cpp:97]: explicit .clear() now removed

File: src/geomgraph/TopologyLocation.cpp (hunk 2 above)
Comment (rouault) [reply src/geomgraph/TopologyLocation.cpp:52]: do we have some ready made benchmark to test this ? all what I can tell is that in RelWithDebInfo mode, the assembly is slightly smaller with the new version.
New version
```
0000000000000000 <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)>:
   0:   f3 0f 1e fa             endbr64 
   4:   0f b6 47 03             movzbl 0x3(%rdi),%eax
   8:   0f b6 56 03             movzbl 0x3(%rsi),%edx
   c:   48 39 c2                cmp    %rax,%rdx
   f:   76 37                   jbe    48 <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)+0x48>
  11:   b8 ff ff ff ff          mov    $0xffffffff,%eax
  16:   48 83 fa 03             cmp    $0x3,%rdx
  1a:   c6 47 03 03             movb   $0x3,0x3(%rdi)
  1e:   66 89 47 01             mov    %ax,0x1(%rdi)
  22:   b8 03 00 00 00          mov    $0x3,%eax
  27:   48 0f 47 d0             cmova  %rax,%rdx
  2b:   31 c0                   xor    %eax,%eax
  2d:   80 3c 07 ff             cmpb   $0xff,(%rdi,%rax,1)
  31:   75 07                   jne    3a <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)+0x3a>
  33:   0f b6 0c 06             movzbl (%rsi,%rax,1),%ecx
  37:   88 0c 07                mov    %cl,(%rdi,%rax,1)
  3a:   48 83 c0 01             add    $0x1,%rax
  3e:   48 39 c2                cmp    %rax,%rdx
  41:   75 ea                   jne    2d <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)+0x2d>
  43:   c3                      retq   
  44:   0f 1f 40 00             nopl   0x0(%rax)
  48:   48 85 d2                test   %rdx,%rdx
  4b:   75 de                   jne    2b <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)+0x2b>
  4d:   c3                      retq   
```
vs original one
```
0000000000000000 <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)>:
   0:   f3 0f 1e fa             endbr64 
   4:   0f b6 57 03             movzbl 0x3(%rdi),%edx
   8:   0f b6 4e 03             movzbl 0x3(%rsi),%ecx
   c:   38 ca                   cmp    %cl,%dl
   e:   73 40                   jae    50 <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)+0x50>
  10:   b8 ff ff ff ff          mov    $0xffffffff,%eax
  15:   c6 47 03 03             movb   $0x3,0x3(%rdi)
  19:   66 89 47 01             mov    %ax,0x1(%rdi)
  1d:   31 c0                   xor    %eax,%eax
  1f:   80 3c 07 ff             cmpb   $0xff,(%rdi,%rax,1)
  23:   75 05                   jne    2a <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)+0x2a>
  25:   48 39 c1                cmp    %rax,%rcx
  28:   77 16                   ja     40 <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)+0x40>
  2a:   0f b6 57 03             movzbl 0x3(%rdi),%edx
  2e:   48 83 c0 01             add    $0x1,%rax
  32:   48 39 c2                cmp    %rax,%rdx
  35:   77 e8                   ja     1f <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)+0x1f>
  37:   c3                      retq   
  38:   0f 1f 84 00 00 00 00 00         nopl   0x0(%rax,%rax,1)
  40:   0f b6 14 06             movzbl (%rsi,%rax,1),%edx
  44:   88 14 07                mov    %dl,(%rdi,%rax,1)
  47:   eb e1                   jmp    2a <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)+0x2a>
  49:   0f 1f 80 00 00 00 00    nopl   0x0(%rax)
  50:   84 d2                   test   %dl,%dl
  52:   75 c9                   jne    1d <geos::geomgraph::TopologyLocation::merge(geos::geomgraph::TopologyLocation const&)+0x1d>
  54:   c3                      retq   
```

Comment (dbaston) [reply src/geomgraph/TopologyLocation.cpp:52]: > do we have some ready made benchmark to test this ? 

Honestly, no. Usually I would call something general-purpose like `perf_intersection` but that no longer hits this code, and it doesn't look like there's much left that does. Maybe only single-sided buffer, and that's probably an oversight. So I'm not going to worry about it.

**Pull request merged**